android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java

```diff
@@ -151,10 +154,54 @@ public void setShouldShowHelpBubble(boolean flag) {
     this._shouldShowHelpBubble = flag;
   }
 
-  protected boolean shouldIgnoreTextChange() { return shouldIgnoreTextChange; }
-  protected void setShouldIgnoreTextChange(boolean ignore) { this.shouldIgnoreTextChange = ignore; }
-  protected boolean shouldIgnoreSelectionChange() { return shouldIgnoreSelectionChange; }
-  protected void setShouldIgnoreSelectionChange(boolean ignore) { this.shouldIgnoreSelectionChange = ignore; }
+  protected boolean shouldIgnoreTextChange() {
+    return shouldIgnoreTextChange;
+  }
+
+  protected void setShouldIgnoreTextChange(boolean ignore) {
+    this.shouldIgnoreTextChange = ignore;
+  }
+
+  protected boolean shouldIgnoreSelectionChange() {
+    return shouldIgnoreSelectionChange;
+  }
+
+  protected void setShouldIgnoreSelectionChange(boolean ignore) {
+    this.shouldIgnoreSelectionChange = ignore;
+  }
+
+  protected boolean updateText(String text) {
+    boolean result = false;
+    String kmText = "";
+    if (text != null) {
+      kmText = text.toString().replace("\\", "\\u005C").replace("'", "\\u0027").replace("\n", "\\n");
+    }
+
+    if (KMManager.isKeyboardLoaded(this.keyboardType) && !shouldIgnoreTextChange) {
+      this.loadJavascript(KMString.format("updateKMText('%s')", kmText));
+      result = true;
+    }
+
+    shouldIgnoreTextChange = false;
+    return result;
+  }
+
+  protected boolean updateSelectionRange(int selStart, int selEnd) {
+    boolean result = false;
+    InputConnection ic = KMManager.getInputConnection(this.keyboardType);
```

Comment:
Just noting this as a future potential item - it may be nice for the `KMKeyboard` class to own/maintain the reference to its own connection.